### PR TITLE
[FIXED JENKINS-47653] Lazy evaluation of branch last modified timestamp for Bitbucket Server

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-47653](https://issues.jenkins-ci.org/browse/JENKINS-47653)

### CODE REVIEW

**Comment by casz** [comment]
sad to not see the java 8 bump 😢

**Comment by stephenc** [comment]
@casz java 8 bump would prevent some users with Jenkins < 2.60 from getting this critical fix

**Comment by stephenc** [comment]
@casz the Java 8 bump will come soon when git 4.0.0 bumps its baseline Jenkins to 2.60 (early in the new year)

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java [hunk 1]
```diff
@@ -98,8 +99,8 @@
     private static final String API_REPOSITORIES_PATH = API_BASE_PATH + "/projects/%s/repos?start=%s";
     private static final String API_REPOSITORY_PATH = API_BASE_PATH + "/projects/%s/repos/%s";
     private static final String API_DEFAULT_BRANCH_PATH = API_BASE_PATH + "/projects/%s/repos/%s/branches/default";
-    private static final String API_BRANCHES_PATH = API_BASE_PATH + "/projects/%s/repos/%s/branches?start=%s";
-    private static final String API_PULL_REQUESTS_PATH = API_BASE_PATH + "/projects/%s/repos/%s/pull-requests?start=%s";
+    private static final String API_BRANCHES_PATH = API_BASE_PATH + "/projects/%s/repos/%s/branches?start=%s&limit=%s";
```
Comment (rsandell) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: shouldn't it be `"limit=%d"` ?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java [hunk 2]
```diff
@@ -253,8 +255,9 @@ public String getRepositoryUri(@NonNull BitbucketRepositoryType type,
                 if (Thread.interrupted()) {
                     throw new InterruptedException();
                 }
+                Integer limit = page.getLimit();
```
Comment (rsandell) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: could this be `<= 0` ? in that case we possibly risk unnecessary looping.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/branch/BitbucketServerBranch.java [hunk 3]
```diff
@@ -80,4 +87,25 @@ public void setTimestamp(long timestamp) {
         this.timestamp = timestamp;
     }
 
+    @Restricted(NoExternalUse.class)
+    public void setTimestampClosure(Callable<Long> timestampClosure) {
+        this.timestampClosure = timestampClosure;
+    }
+
+    private long timestamp() {
```
Comment (rsandell) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/branch/BitbucketServerBranch.java]: Not thread safe. Do you need that?

Comment (stephenc) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/branch/BitbucketServerBranch.java]: no need for thread-safe... worst case it should be idempotent

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java (hunk 2 above)
Comment (stephenc) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: well if less than 0 bitbucket server is playing silly monkeys

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java (hunk 1 above)
Comment (stephenc) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: I'll fix-up post merge

Comment (stephenc) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: c1ac1c9828239b2b4444b0e9ddbefdcdd04028d8

**Pull request merged**